File: MapReduce_FastQ/src/main/java/PerBaseSequenceContent/FastQ_PerBaseSequenceContent_Mapper.java
package PerBaseSequenceContent;

import org.apache.hadoop.io.IntWritable;
import org.apache.hadoop.io.Text;
import org.seqdoop.hadoop_bam.SequencedFragment;

import java.io.IOException;

/**
 * master-thesis Clemens Banas
 * Organization: DBIS - University of Innsbruck
 * Created 06.08.15.
 */
public class FastQ_PerBaseSequenceContent_Mapper extends org.apache.hadoop.mapreduce.Mapper<Object, SequencedFragment, IntWritable, Text> {
    private static final char BASE_A = 'A';
    private static final char BASE_C = 'C';
    private static final char BASE_G = 'G';
    private static final char BASE_T = 'T';

    @Override
    protected void map(Object key, SequencedFragment value, Context context) throws IOException, InterruptedException {
        final Text sequence = value.getSequence();
        for (int i = 0; i < sequence.getLength(); i++) {
            IntWritable outputKey = new IntWritable(i+1);
            switch (sequence.charAt(i)) {
                case BASE_A:
                    context.write(outputKey, new Text(String.valueOf(BASE_A)));
                    break;
                case BASE_C:
                    context.write(outputKey, new Text(String.valueOf(BASE_C)));
                    break;
                case BASE_G:
                    context.write(outputKey, new Text(String.valueOf(BASE_G)));
                    break;
                case BASE_T:
                    context.write(outputKey, new Text(String.valueOf(BASE_T)));
                    break;
                default:
                    break;
            }
        }
    }

}

File: Spark_FastQ_Scala/src/main/scala/PerBaseSequenceContent/FastQ_PerBaseSequenceContent_Mapper.scala
package main.scala.PerBaseSequenceContent

import org.apache.hadoop.io.Text
import org.seqdoop.hadoop_bam.SequencedFragment

import scala.collection.mutable.ListBuffer

/**
  * master-thesis Clemens Banas
  * Organization: DBIS - University of Innsbruck
  * Created 02.01.16.
  */
object FastQ_PerBaseSequenceContent_Mapper {
  private val BASE_A: Char = 'A'
  private val BASE_C: Char = 'C'
  private val BASE_G: Char = 'G'
  private val BASE_T: Char = 'T'

  def flatMap(record: SequencedFragment): TraversableOnce[Pair[Int,Char]] = {
    val resList = new ListBuffer[Pair[Int,Char]]()
    val sequence : Text = record.getSequence()
    for( i <- 0 until sequence.getLength ) {
      val outputKey = i+1
      sequence.charAt(i) match {
        case BASE_A => resList.append(new Pair(outputKey, BASE_A))
        case BASE_C => resList.append(new Pair(outputKey, BASE_C))
        case BASE_G => resList.append(new Pair(outputKey, BASE_G))
        case BASE_T => resList.append(new Pair(outputKey, BASE_T))
        case default => println("base character '" + default + "' occurred at position " + outputKey)
      }
    }
    resList.toTraversable
  }

}
